### PR TITLE
Java.Util.ArrayList

### PR DESCRIPTION
**Pull request by frangte**

### CODE REVIEW

**Comment by thoangminh** [comment]
Please check my comment in channel #review. :)

**Review by ghost — CHANGES_REQUESTED** [review]
Please check my comment in channel #review. :)

File: java/util/ArrayList/ArrayList.hpp [hunk 1]
```diff
@@ -194,7 +171,7 @@ namespace Java {
 			 * @return boolean
 			 */
 			boolean addAll(Collection<E> collection) {
-				// TODO
+				// TODO: Please discuss about param Collection
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 2]
```diff
@@ -203,11 +180,11 @@ namespace Java {
 			 * starting at the specified position.
 			 *
 			 * @param index
-			 * @param c
+			 * @param collection
 			 * @return boolean
 			 */
 			boolean addAll(int index, Collection<E> collection) {
-				// TODO
+				// TODO: Please discuss about param Collection
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 3]
```diff
@@ -222,31 +199,35 @@ namespace Java {
...
 			/**
+			 * Returns true if this collection contains all of the elements in the specified collection.
 			 *
-			 * @param c
-			 * @return
+			 * @param collection
+			 * @return boolean
 			 */
 			boolean containsAll(Collection<Object> collection) const {
+				// TODO: Please discuss about param Collection
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 4]
```diff
@@ -268,18 +249,20 @@ namespace Java {
 			 * @param action
 			 */
 			void forEach(Consumer<E> action) const {
-				// TODO
+                // TODO: Please complete class Consumer
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 5]
```diff
@@ -310,7 +299,7 @@ namespace Java {
 			 * @return Address of Iterator<E>
 			 */
 			Iterator<E> &iterator() const {
-				// TODO
+                // TODO: Please complete class Iterator
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 6]
```diff
@@ -382,7 +385,7 @@ namespace Java {
 			 * @return boolean
 			 */
 			boolean removeAll(Collection<Object> target) {
-				// TODO
+                // TODO: Please discuss about param Collection
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 7]
```diff
@@ -393,7 +396,7 @@ namespace Java {
 			 * @return boolean
 			 */
 			boolean removeIf(Predicate<E> filter) {
-				// TODO
+                // TODO: Please complete class Predicate
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 8]
```diff
@@ -404,18 +407,18 @@ namespace Java {
 			 * @param unaryOperator
 			 */
 			void replaceAll(UnaryOperator<E> unaryOperator) {
-				// TODO
+                // TODO: Please complete class UnaryOperator
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 9]
```diff
@@ -404,18 +407,18 @@ namespace Java {
...
 			 * in the specified collection.
 			 *
-			 * @param c
-			 * @return
+			 * @param collection
+			 * @return boolean
 			 */
 			boolean retainAll(Collection<Object> collection) {
-				// TODO
+                // TODO: Please discuss about Param Collection
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 10]
```diff
@@ -447,7 +455,7 @@ namespace Java {
 			 * @param cmp
 			 */
 			void sort(Comparator<E> cmp) {
-				// TODO
+                // TODO: Please complete class Comparator
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 11]
```diff
@@ -456,13 +464,13 @@ namespace Java {
 			 * @return Spliterator<E>
 			 */
 			Spliterator<E> &spliterator() const {
-				// TODO
+                // TODO: Please complete class Spliterator
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 12]
```diff
@@ -517,35 +525,55 @@ namespace Java {
...
+			 *
+			 * @return long
+			 */
 			long hashCode() const {
-				return 0;
+				return Object::hashCode();
 			}
-			
+
+            // TODO: Please discuss about param Object
```
Comment (thoangminh) [inline java/util/ArrayList/ArrayList.hpp]: You can replace TODO by TODO(your name) to let people know this is your TODO.

File: java/util/ArrayList/ArrayList.hpp [hunk 13]
```diff
@@ -556,7 +584,7 @@ namespace Java {
 			 * @param toIndex
 			 */
 			void removeRange(int fromIndex, int toIndex) {
-				// TODO
+				this->original.erase(this->original.begin() + fromIndex, this->begin() + toIndex);
```
Comment (ghost) [inline java/util/ArrayList/ArrayList.hpp]: Please plit or shorten this line. Your code style will be better :)

**Review by ghost — DISMISSED** [review]

**Review by dthongvl — APPROVED** [review]

**Pull request merged**